zeroex/orderwatch/order_watcher.go

```diff
@@ -1766,33 +1818,54 @@ func (w *Watcher) removeAssetDataAddressFromEventDecoder(assetData []byte) error
 }
 
 func (w *Watcher) increaseMaxExpirationTimeIfPossible() error {
-	if orderCount, err := w.db.CountOrders(nil); err != nil {
+	orderCount, err := w.db.CountOrders(nil)
+	if err != nil {
 		return err
-	} else if orderCount < w.maxOrders {
+	}
+	if orderCount < w.maxOrders {
 		// We have enough space for new orders. Set the new max expiration time to the
 		// value of slow counter.
 		newMaxExpiration := w.maxExpirationCounter.Count()
-		if w.maxExpirationTime.Cmp(newMaxExpiration) != 0 {
+		if newMaxExpiration.Cmp(w.maxExpirationTime) == 1 {
 			logger.WithFields(logger.Fields{
 				"oldMaxExpirationTime": w.maxExpirationTime.String(),
 				"newMaxExpirationTime": fmt.Sprint(newMaxExpiration),
 			}).Debug("increasing max expiration time")
-			w.maxExpirationTime.Set(newMaxExpiration)
-			w.saveMaxExpirationTime(newMaxExpiration)
+			w.setMaxExpirationTime(newMaxExpiration)
+		}
+	}
+
+	// The minimum max expiration time should be the latest block timestamp + 1.
```

Comment:
This change is a departure from the previous implementation of dynamic max expiration time. I don't think it ever makes sense to have the max expiration time be lower than the current block timestamp. There are some cases where this can happen while we are waiting for the slow counter to catch up.